File: src/week7/P011.js
const fs = require('fs');
const input = fs.readFileSync('/dev/stdin').toString().trim().split('\n');
const [x1, y1, x2, y2] = input[0].trim().split(/\s+/).map(Number);
const [x3, y3, x4, y4] = input[1].trim().split(/\s+/).map(Number);

function ccw(x1, y1, x2, y2, x3, y3) {
    // 좌표값이 매우 클 경우 BigInt를 써야 하지만, 문제 조건(100만 이하)에서는
    // 일반 연산 후 부호만 판별해도 무방합니다.
    const result = (x2 - x1) * (y3 - y1) - (y2 - y1) * (x3 - x1);
    
    if (result > 0) return 1;
    if (result < 0) return -1;
    return 0;
}

function solution() {
    const ccw123 = ccw(x1, y1, x2, y2, x3, y3);
    const ccw124 = ccw(x1, y1, x2, y2, x4, y4);
    const ccw341 = ccw(x3, y3, x4, y4, x1, y1);
    const ccw342 = ccw(x3, y3, x4, y4, x2, y2);

    if (ccw123 * ccw124 === 0 && ccw341 * ccw342 === 0) {
        // Bounding Box 비교 (겹치는지 확인)
        const mx1 = Math.min(x1, x2);
        const my1 = Math.min(y1, y2);
        const mx2 = Math.max(x1, x2);
        const my2 = Math.max(y1, y2);

        const mx3 = Math.min(x3, x4);
        const my3 = Math.min(y3, y4);
        const mx4 = Math.max(x3, x4);
        const my4 = Math.max(y3, y4);

        if (mx1 <= mx4 && mx3 <= mx2 && my1 <= my4 && my3 <= my2) {
            return 1;
        }
    } 
    else {
        if (ccw123 * ccw124 <= 0 && ccw341 * ccw342 <= 0) {
            return 1;
        }
    }
    return 0;
}

console.log(solution());

File: src/week7/P011.java
package week7;

import java.io.*;
import java.util.*;

public class P011 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        long x1 = Long.parseLong(st.nextToken());
        long y1 = Long.parseLong(st.nextToken());
        long x2 = Long.parseLong(st.nextToken());
        long y2 = Long.parseLong(st.nextToken());

        st = new StringTokenizer(br.readLine());
        // 선분 2 (x3, y3, x4, y4)
        long x3 = Long.parseLong(st.nextToken());
        long y3 = Long.parseLong(st.nextToken());
        long x4 = Long.parseLong(st.nextToken());
        long y4 = Long.parseLong(st.nextToken());

        System.out.println(solution(x1, y1, x2, y2, x3, y3, x4, y4));
    }
    static int ccw(long x1, long y1, long x2, long y2, long x3, long y3) {
        long result = (x2 - x1) * (y3 - y1) - (y2 - y1) * (x3 - x1);
        if (result > 0) return 1;   // 반시계
        if (result < 0) return -1;  // 시계
        return 0;                   // 일직선
    }
    static int solution(long x1, long y1, long x2, long y2, long x3, long y3, long x4, long y4) {
        int ccw123 = ccw(x1, y1, x2, y2, x3, y3);
        int ccw124 = ccw(x1, y1, x2, y2, x4, y4);
        int ccw341 = ccw(x3, y3, x4, y4, x1, y1);
        int ccw342 = ccw(x3, y3, x4, y4, x2, y2);
        boolean isStraight = (ccw123 * ccw124 == 0) && (ccw341 * ccw342 == 0);

        if (isStraight) {
            // 포개짐 여부 확인 (Bounding Box Check)
            // 각 선분의 양 끝점을 정렬하여 min, max를 구함
            if (Math.min(x1, x2) <= Math.max(x3, x4) &&
                    Math.min(x3, x4) <= Math.max(x1, x2) &&
                    Math.min(y1, y2) <= Math.max(y3, y4) &&
                    Math.min(y3, y4) <= Math.max(y1, y2)) {
                return 1;
            }
        }
        else {
            // 등호(<=)가 들어가는 이유는 끝점에서 만나는 경우도 교차로 치기 때문
            if (ccw123 * ccw124 <= 0 && ccw341 * ccw342 <= 0) {
                return 1;
            }
        }
        return 0;
    }
}